lib/smppex/pdu/validity_period.ex

```diff
@@ -0,0 +1,89 @@
+defmodule SMPPEX.Pdu.ValidityPeriod do
+  @moduledoc """
+  Module for converting validity period to time diff in seconds.
+  """
+
+  require Logger
+
+  # seconds
+  @max_validity_period 48 * 3600
+  @default_validity_period 24 * 3600
+
+  @spec normalized(String.t(), non_neg_integer) :: non_neg_integer
+
+  @doc """
+  Converts validity_period field to time diff in seconds according to SMPP specification.
+
+  ## Example
+
+      iex> seconds = "2017-01-01 00:00:00" |> NaiveDateTime.from_iso8601!() |> DateTime.from_naive!("Etc/UTC") |> DateTime.to_unix()
+      1483228800
+
+      iex> SMPPEX.Pdu.ValidityPeriod.normalized("170102233429004+", seconds)
+      167669
+
+  """
+
+  def normalized(validity_period, time_now \\ System.system_time(:second))
+
+  def normalized(nil, _) do
+    @default_validity_period
+  end
+
+  def normalized("", _) do
+    @default_validity_period
+  end
+
+  def normalized(validity_period, time_now) do
+    seconds = to_seconds(validity_period, time_now)
+
+    if 0 < seconds && seconds <= @max_validity_period do
+      seconds
+    else
+      @default_validity_period
+    end
+  rescue
+    e ->
+      Logger.warn("Invalid validity_period #{inspect(validity_period)}:#{inspect(e)}")
+      @default_validity_period
+  end
+
+  @spec to_seconds(String.t(), non_neg_integer) :: non_neg_integer
+  def to_seconds(
+        <<y::binary-size(2), m::binary-size(2), d::binary-size(2), h::binary-size(2),
+          mn::binary-size(2), s::binary-size(2), _t::binary-size(1), _nn::binary-size(2), "R">>,
+        _
+      ) do
+    String.to_integer(s) +
+      String.to_integer(mn) * 60 +
+      String.to_integer(h) * 3600 +
+      String.to_integer(d) * 24 * 3600 +
+      String.to_integer(m) * 30 * 24 * 3600 +
+      String.to_integer(y) * 12 * 30 * 24 * 3600
```

Comment:
We have to choose whether we want to use this naive year/month shift implementation or we should include some third-party dependency for dealing with time and date shifts (Timex/edate/qdate etc)